components/capi/src/pluralrules.rs

```diff
@@ -0,0 +1,129 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use icu_locid::Locale as ICULocale;
+use icu_plurals::PluralCategory as ICUPluralCategory;
+use icu_plurals::PluralOperands as ICUPluralOperands;
+use icu_plurals::PluralRuleType as ICUPluralRuleType;
+
+use crate::provider::ErasedDataProvider;
+
+use std::ptr;
+
+// opaque type
+pub use icu_plurals::PluralRules;
+
+#[repr(C)]
+pub struct CreatePluralRulesResult {
+    rules: *mut PluralRules,
+    success: bool,
+}
+
+#[no_mangle]
+pub extern "C" fn plural_rules_create(
+    locale: &ICULocale,
+    provider: &mut ErasedDataProvider,
+    ty: PluralRuleType,
+) -> CreatePluralRulesResult {
+    // cheap as long as there are no variants
+    let langid = locale.as_ref().clone();
+    let provider = provider.as_dyn_ref();
+    match PluralRules::try_new(langid, provider, ty.into()) {
+        Ok(pr) => {
+            let pr = Box::new(pr);
+            CreatePluralRulesResult {
+                rules: Box::into_raw(pr),
+                success: true,
+            }
+        }
+        Err(_) => CreatePluralRulesResult {
+            rules: ptr::null_mut(),
+            success: false,
+        },
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn plural_rules_select(pr: &PluralRules, op: PluralOperands) -> PluralCategory {
+    pr.select(op).into()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn plural_rules_destroy(pr: *mut PluralRules) {
+    let _ = Box::from_raw(pr);
+}
+
+#[repr(C)]
+pub struct PluralOperands {
+    pub i: u64,
+    pub v: usize,
+    pub w: usize,
+    pub f: u64,
+    pub t: u64,
+    pub c: usize,
+}
```

Comment:
Are you duplicating the enums only temporarily, or do you intend that we should always duplicate them instead of adding `#[repr(C)]` to the main types?

Reply:
I prefer permanent duplication, the conversions below will stop compiling if you make incompatible changes, and this allows us to potentially change the Rust side while keeping the C side compatible.

My design goal here is that everything relevant to FFI is _in this crate itself_ so people outside this crate don't need to care about it.

Reply:
Longer-term, do you envision auto-generating the FFI enums from Rust enums, or is it by design that you need to update it in both places (to make FFI changes more explicit)?

How do we deal with API docs if we have duplicated structs?  If the Rust struct has API docs, what happens with the FFI struct?

Reply:
> Longer-term, do you envision auto-generating the FFI enums from Rust enums, or is it by design that you need to update it in both places (to make FFI changes more explicit)?

By design, IMO. I talk about this in the "bridge blocks are self-contained" section of the [`diplomat` doc](https://docs.google.com/document/d/1n0Igz_q6uUm9OqlRgWJfmtlgdeJ1lh6ubP1rRIIHDYM/edit?resourcekey=0-weURPKqKrEvyQc9qNPYCcg#heading=h.b3rpgbhignh0). `cxx` has this restriction and I think it's a good thing, requiring your tooling to understand your crate graph can be fraught and subject to things moving out from underneath you.

The API docs are a good point. I don't know what to do here, we could have manual comments asking people to update things cross file, but that's not great. We could potentially teach tooling to scrape docs (and only docs) from other parts of the codebase, but I'm super wary of scraping anything other than docs.

Reply:
We should come up with a story for docs soon, but we don't need it right now.  Lack of documentation is a key paint point for ICU and ICU wrappers like PyICU.

Maybe something along the lines of `#[doc(include = "...")]` could help?

https://doc.rust-lang.org/unstable-book/language-features/external-doc.html

Reply:
I think we should at least have a hyperlink back to the main Rust type from the docstring.

Really, the place where we need the docs is in the `.h` files, and we can't use `rustdoc` there anyway.  We'll need to come up with a solution for documenting the C header files.

Reply:
Yeah I would just hyperlink back tbh

I think some of the automated tools can copy docs over as comments (perhaps in ways pandoc will like?)

Reply:
I'm a little nervous about letting the C representation drift too far from the Rust structures. There is a Rust library in Firefox with a heavyweight FFI layer that maps between internal Rust types and FFI Rust types that are exposed to C. That has been difficult to test properly, and I think we've had more bugs in that FFI layer than in the Rust or C++ code on either side of it.

Reply:
It won't drift unless intentionally; due to the `From` implementation we are essentially checking that they always have identical fields.

Reply:
It was the intentional drift that I was worried about :)